airbyte-integrations/connectors/source-asana/source_asana/streams.py

```diff
@@ -0,0 +1,241 @@
+    """
+
+    def read_records(
+        self,
+        sync_mode: SyncMode,
+        cursor_field: List[str] = None,
+        stream_slice: Mapping[str, Any] = None,
+        stream_state: Mapping[str, Any] = None,
+    ) -> Iterable[Mapping[str, Any]]:
+        workspaces_stream = Workspaces(authenticator=self.authenticator)
+        for workspace in workspaces_stream.read_records(sync_mode=SyncMode.full_refresh):
+            stream_state = stream_state or dict()
+            pagination_complete = False
+
+            next_page_token = None
+            while not pagination_complete:
+                request_headers = self.request_headers(
+                    stream_state=stream_state, stream_slice=stream_slice, next_page_token=next_page_token
+                )
+                request_params = self.request_params(stream_state=stream_state, stream_slice=stream_slice, next_page_token=next_page_token)
+                request = self._create_prepared_request(
+                    path=self.path(stream_state=stream_state, stream_slice=stream_slice, next_page_token=next_page_token),
+                    headers=dict(request_headers, **self.authenticator.get_auth_header()),
+                    params=dict(request_params, workspace=workspace["gid"]),
+                    json=self.request_body_json(stream_state=stream_state, stream_slice=stream_slice, next_page_token=next_page_token),
+                )
+
+                response = self._send_request(request)
+                yield from self.parse_response(response, stream_state=stream_state, stream_slice=stream_slice)
+
+                next_page_token = self.next_page_token(response)
+                if not next_page_token:
+                    pagination_complete = True
+
+        # Always return an empty generator just in case no records were ever yielded
+        yield from []
+
+
+class CustomFields(AsanaStream):
+    def path(self, stream_slice: Mapping[str, Any] = None, **kwargs) -> str:
+        workspace_gid = stream_slice["workspace_gid"]
+        return f"workspaces/{workspace_gid}/custom_fields"
+
+    def read_records(self, stream_slice: Optional[Mapping[str, Any]] = None, **kwargs) -> Iterable[Mapping[str, Any]]:
+        workspaces_stream = Workspaces(authenticator=self.authenticator)
+        for workspace in workspaces_stream.read_records(sync_mode=SyncMode.full_refresh):
+            yield from super().read_records(stream_slice={"workspace_gid": workspace["gid"]}, **kwargs)
+
+
+class Projects(AsanaStreamWorkspacePagination):
+    def path(self, **kwargs) -> str:
+        return "projects"
+
+
+class Sections(AsanaStream):
+    def path(self, stream_slice: Mapping[str, Any] = None, **kwargs) -> str:
+        project_gid = stream_slice["project_gid"]
+        return f"projects/{project_gid}/sections"
+
+    def read_records(self, stream_slice: Optional[Mapping[str, Any]] = None, **kwargs) -> Iterable[Mapping[str, Any]]:
```

Comment:
better use stream_slices and remove read_records override

Reply:
Done.